File: javatest/omnia/data/cache/SimpleDoubleMemoizerTest.java
package omnia.data.cache;

import static com.google.common.truth.Truth.assertThat;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.times;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.when;

import java.util.function.DoubleSupplier;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.junit.runners.JUnit4;

@RunWith(JUnit4.class)
public class SimpleDoubleMemoizerTest {

  @Test
  public void value_didReturnSuppliedResult() {
    double testValue = 132.0;
    MemoizedDouble testSubject = new SimpleDoubleMemoizer(() -> testValue);

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void value_twice_didReturnSuppliedResult() {
    double testValue = 132.0;
    MemoizedDouble testSubject = new SimpleDoubleMemoizer(() -> testValue);

    testSubject.value();

    assertThat(testSubject.value()).isEqualTo(testValue);
  }

  @Test
  public void value_didInvokeSupplier() {
    DoubleSupplier supplier = setUpSupplier();
    MemoizedDouble testSubject = new SimpleDoubleMemoizer(supplier);

    testSubject.value();

    verify(supplier).getAsDouble();
  }

  @Test
  public void new_didNotInvokeSupplier() {
    DoubleSupplier supplier = setUpSupplier();
    new SimpleDoubleMemoizer(supplier);

    verify(supplier, never()).getAsDouble();
  }

  @Test
  public void value_twice_onlyInvokedSupplierOnce() {
    DoubleSupplier supplier = setUpSupplier();
    MemoizedDouble testSubject = new SimpleDoubleMemoizer(supplier);

    testSubject.value();
    testSubject.value();

    verify(supplier, times(1)).getAsDouble();
  }

  @Test
  public void new_withNullSupplier_didThrowException() {
    assertThrows(NullPointerException.class, () -> new SimpleDoubleMemoizer(null));
  }

  private static DoubleSupplier setUpSupplier() {
    DoubleSupplier supplier = mock(DoubleSupplier.class);
    when(supplier.getAsDouble()).thenReturn(132.0);
    return supplier;
  }
}

File: tests/omnia/data/cache/SimpleDoubleMemoizerTest.kt
package omnia.data.cache

import com.google.common.truth.Truth.assertThat
import java.util.function.DoubleSupplier
import kotlin.test.Test

import org.mockito.Mockito

class SimpleDoubleMemoizerTest {

  @Test
  fun value_didReturnSuppliedResult() {
    val testValue = 132.0
    val testSubject: MemoizedDouble = SimpleDoubleMemoizer { testValue }
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun value_twice_didReturnSuppliedResult() {
    val testValue = 132.0
    val testSubject: MemoizedDouble = SimpleDoubleMemoizer { testValue }
    testSubject.value()
    assertThat(testSubject.value()).isEqualTo(testValue)
  }

  @Test
  fun value_didInvokeSupplier() {
    val supplier = setUpSupplier()
    val testSubject: MemoizedDouble = SimpleDoubleMemoizer(supplier)
    testSubject.value()
    Mockito.verify(supplier).asDouble
  }

  @Test
  fun new_didNotInvokeSupplier() {
    val supplier = setUpSupplier()
    SimpleDoubleMemoizer(supplier)
    Mockito.verify(supplier, Mockito.never()).asDouble
  }

  @Test
  fun value_twice_onlyInvokedSupplierOnce() {
    val supplier = setUpSupplier()
    val testSubject: MemoizedDouble = SimpleDoubleMemoizer(supplier)
    testSubject.value()
    testSubject.value()
    Mockito.verify(supplier, Mockito.times(1)).asDouble
  }

  companion object {

    private fun setUpSupplier(): DoubleSupplier {
      val supplier = Mockito.mock(DoubleSupplier::class.java)
      Mockito.`when`(supplier.asDouble).thenReturn(132.0)
      return supplier
    }
  }
}